### ISSUE TITLE
AI: more sensible use of ammo?

### ISSUE THREAD

**Issue by lheckemann** [endless-sky/endless-sky#2349]
Currently, a Navy Cruiser will unload its full armament on a pirate flying a Sparrow. This seems like a huge waste of ammo, considering that the torpedoes will almost never reach their mark, as the smaller and faster missiles will have eliminated the pirate long ago.

**Comment by KiLEdEnNis** [comment]
giving them the frugal tag could help, but in that case Rainmakers wouldn't do anything if they outnumber the enemy
it may not be an issue, but full missile ships should kinda shoot themselves as well

**Comment by gunqqer** [comment]
Yeah missile usage is terrible. I have to micro manage my fleets usage so I don't spend enough ammo to kill a world ship to kill a chaser

**Comment by VonZorch** [comment]
Escorts on frugal almost never fire missiles and on always will flush every launcher at the smallest target.
I'm now selling any secondary weapons on escorts and replacing with guns,

**Comment by gunqqer** [comment]
I never used nor will ever use probably, but when dicking around I played with finisher missiles and model needle missile. Those needles were OP, 12 on a shield beetle (8 guns 4 turrets) and each did like 1.2k armour dps with 10% bleed through. Shot like 4 per second and when they killed a korsth raider they would shoot a massive amount at the chasers. Wasted so much

**Comment by tehhowch** [comment]
The `always` preferences needs a filter on ship class (and then a new name (`smart`?) since always means *always* ).
Some `smart` preference possibilities:
- `always` behavior until some fraction of ammo is remaining (50-75%)
- after 50-75% ammo capacity is reached, `beFrugal` if target ship is a fighter, drone, or interceptor (etc)
- frugal's "we're outgunned" behavior does not consider the whole system, but only within a given radius of the ship.

**Comment by tehhowch** [comment]
Another missing part of ammo use that I wouldn't mind adding is consideration of asteroid flux to the target-in-range/findtarget routines. That would help a bit with nonsensical ammo usage, especially for non safe blast radius weapons.

**Comment by VonZorch** [comment]
Also, missiles with a blast radius should not fired while the range to target is within the blast radius.
Or those weapons should be changed so that the damage in the weapon definition is much lower and not have a blast radius with a range that just clears the blast radius, then have a sub-munition with the full effect of the original weapon.  This would give am arming distance for such weapons.

**Comment by Pointedstick** [comment]
Potential considerations:
- Don't fire missiles when currently in-flight missiles would already deal more than, say, 125% or 150% of the necessary damage to destroy the target if they all hit
- Don't fire missiles with less than a certain turn speed at fighters and interceptors
- If the firing ship has an outfit or tactical scanner, calculate the missile weight needed to oversaturate a target's anti-missile defenses and only fire missiles when fewer than 100% of them would be destroyed before they impact
- With a new personality trait (`reckless`, `stupid`, etc), override all the above conditions if evaluating them to false would mean that a ship can't shoot anything (e.g. because the only armament is missile weapons)
- Consider missiles fired by allies for the purpose of all the above calculations

**Comment by Tadrix** [comment]
> but only within a given radius of the ship

Three things that ship could (should?) consider:
— Who's in my missile range?
— How many ships have me in their weapons range?
— How much health I have left?

Being in a target-rich environment and being in danger changes behaviour (with a good reason), could use the same logic in the game. Shields or hull below 50%? Run like hell and expend ammo as if there's no tomorrow (because it might not come). I'm targetable by too many to list? Could be a problem. Etc.

**Comment by Amazinite** [comment]
#4224 is related to this.
